File: src/core/validate.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Sukesh Ashok Kumar and The MCDF Project
#include "mcdf/core/validate.hpp"

#include <set>
#include <string>

#include "mcdf/container/container.hpp"
#include "mcdf/core/manifest_ops.hpp"
#include "mcdf/core/render.hpp"
#include "mcdf/core/sealed.hpp"
#include "mcdf/core/sign_ops.hpp"
#include "mcdf/crypto/enc_keys.hpp"
#include "mcdf/model/document.hpp"
#include "mcdf/serialize/policy_yaml.hpp"

namespace mcdf {
namespace {

bool profile_at_least(Profile p, Profile floor) {
  return static_cast<int>(p) >= static_cast<int>(floor);
}

void check_core(const Container& c, const Document& d, ValidationReport& r) {
  if (!d.has_content) {
    r.issues.push_back({"E_MISSING_CONTENT", "content.md is missing"});
  }

  if (!d.has_schema) return;  // no schema, no structural claims to check

  // The set of section ids that bind. Normally that comes from the headings; a
  // sealed content.md has none a validator can read, so spec 5.2.1 puts the
  // same information in the policy and the rule below does not change.
  std::set<std::string> bound;
  const bool content_sealed = d.is_sealed("content.md");

  if (content_sealed) {
    std::string raw_policy;
    if (auto raw = c.read("encryption/policy.yaml"); raw) raw_policy = *raw;
    auto policy = parse_encryption_policy_yaml(raw_policy);
    if (!policy || !policy->structure.present) {
      // Without the attestation the binding is simply not knowable. Saying so
      // once is right; reporting every required section as "missing" would
      // blame the document for the encryptor's omission.
      r.issues.push_back(
          {"E_POLICY_INVALID",
           "content.md is sealed but encryption/policy.yaml carries no structure "
           "attestation, so schema binding cannot be evaluated (spec 5.2.1)"});
      return;
    }
    bound.insert(policy->structure.bound_sections.begin(),
                 policy->structure.bound_sections.end());
  } else {
    for (const auto& h : d.headings) {
      if (!h.id.empty()) bound.insert(h.id);
    }
  }

  const std::string where = content_sealed ? "in the sealed content" : "in content";
  for (const auto& s : d.schema.sections) {
    if (s.id.empty() || bound.contains(s.id)) continue;
    if (s.required) {
      r.issues.push_back(
          {"E_REQUIRED_SECTION_MISSING",
           "required section has no matching heading " + where + ": " + s.id});
    } else {
      r.issues.push_back(
          {"E_SCHEMA_UNBOUND",
           "schema section has no matching heading " + where + ": " + s.id});
    }
  }
}

void check_integrity(const Container& c, const Document& d,
                     ValidationReport& r) {
  if (!d.has_manifest) {
    r.issues.push_back({"E_MISSING_MANIFEST", "manifest.json is missing"});
    return;
  }
  if (d.manifest.hash_algorithm != kDefaultHashAlgorithm) {
    r.issues.push_back({"E_ALGO_NOT_ALLOWED",
                        "hash algorithm not allowed: " +
                            d.manifest.hash_algorithm});
    return;
  }

  auto v = verify_manifest(c, d.manifest);
  if (!v) {
    r.issues.push_back({"E_IO", v.error().message});
    return;
  }
  for (const auto& p : v->missing)
    r.issues.push_back(
        {"E_MANIFEST_MISSING_FILE", "file listed in manifest is missing: " + p});
  for (const auto& p : v->mismatched)
    r.issues.push_back({"E_MANIFEST_HASH_MISMATCH", "content hash mismatch: " + p});
  for (const auto& p : v->extra)
    r.issues.push_back(
        {"E_MANIFEST_EXTRA_FILE", "file present but not listed in manifest: " + p});
}

// Verifies every signatures/*.sig over the canonical manifest, mapping each
// failure to its normative code. Classification keys off this library's own
// verifier messages (same translation unit family, kept in sync).
void check_signed(const Container& c, ValidationReport& r) {
  auto checks = verify_container(c);
  if (!checks) {
    r.issues.push_back({"E_IO", checks.error().message});
    return;
  }
  if (checks->empty()) {
    r.issues.push_back(
        {"E_SIG_MISSING", "no signatures present (signatures/*.sig)"});
    return;
  }
  for (const auto& s : *checks) {
    if (s.valid) continue;
    const std::string& e = s.error;
    if (e.find("unsupported JWS alg") != std::string::npos) {
      r.issues.push_back({"E_SIG_ALG_NOT_ALLOWED", s.file + ": " + e});
    } else if (e.find("does not match key type") != std::string::npos) {
      r.issues.push_back({"E_SIG_ALG_MISMATCH", s.file + ": " + e});
    } else if (e.find("did:key") != std::string::npos ||
               e.find("did key") != std::string::npos) {
      r.issues.push_back({"E_KID_UNRESOLVABLE", s.file + ": " + e});
    } else {
      r.issues.push_back(
          {"E_SIG_INVALID",
           s.file + ": " +
               (e.empty() ? "signature does not verify over the canonical manifest"
                          : e)});
    }
  }
}

// Static soundness of the encryption layer (no key available here, so the
// decrypt-time codes E_NOT_A_RECIPIENT / E_DECRYPT_FAILED stay out of scope).
// An unencrypted container has nothing to assert and passes.
void check_encrypted(const Container& c, ValidationReport& r) {
  if (!c.contains("encryption/policy.yaml")) return;
  auto raw = c.read("encryption/policy.yaml");
  if (!raw) {
    r.issues.push_back({"E_IO", raw.error().message});
    return;
  }
  auto p = parse_encryption_policy_yaml(*raw);
  if (!p) {
    r.issues.push_back({"E_POLICY_INVALID", p.error().message});
    return;
  }
  if (p->method != "aes-256-gcm")
    r.issues.push_back(
        {"E_ALGO_NOT_ALLOWED", "encryption method not allowed: " + p->method});
  if (p->key_management != "hpke")
    r.issues.push_back({"E_ALGO_NOT_ALLOWED",
                        "key management not allowed: " + p->key_management});
  if (p->encrypted_files.empty())
    r.issues.push_back({"E_POLICY_INVALID", "policy lists no encrypted files"});
  for (const auto& f : p->encrypted_files) {
    if (!c.contains(f))
      r.issues.push_back(
          {"E_POLICY_INVALID",
           "encrypted file listed in policy is missing: " + f});
  }
  if (p->recipients.empty())
    r.issues.push_back({"E_POLICY_INVALID", "policy has no recipients"});

  // The attestation's *presence* rules (spec 5.2.1). The "required but absent"
  // case is reported by check_core, which runs at every profile; what is left
  // here is the reverse — an attestation nobody can corroborate, which either
  // leaks the schema or claims something about content that is not sealed.
  if (p->structure.present) {
    const auto sealed = sealed_members(c);
    if (!sealed.contains("content.md")) {
      r.issues.push_back({"E_POLICY_INVALID",
                          "policy carries a structure attestation but content.md "
                          "is not encrypted"});
    } else if (sealed.contains("schema.yaml") || !c.contains("schema.yaml")) {
      r.issues.push_back(
          {"E_POLICY_INVALID",
           "policy carries a structure attestation with no readable schema.yaml "
           "to corroborate it (spec 5.2.1)"});
    }
  }
  for (const auto& rec : p->recipients) {
    if (rec.id.empty() || rec.enc.empty() || rec.wrapped_key.empty()) {
      r.issues.push_back({"E_POLICY_INVALID", "recipient entry is incomplete"});
      continue;
    }
    if (!EncPublicKey::from_did_key(rec.id))
      r.issues.push_back(
          {"E_KID_UNRESOLVABLE",
           "recipient id is not a resolvable X25519 did:key: " + rec.id});
  }
}

// The canonical renders must be producible — except over sealed content, which
// cannot be rendered at all, so the profile is not applicable rather than
// failed (spec 6).
void check_render(const Container& c, const Document& d, ValidationReport& r) {
  if (d.is_sealed("content.md")) return;
  for (RenderFormat f : {RenderFormat::kHtml, RenderFormat::kText}) {
    if (auto out = render(c, f); !out)
      r.issues.push_back({"E_IO", "render " + std::string(to_string(f)) +
                                      " failed: " + out.error().message});
  }
}

}  // namespace

Result<Profile> parse_profile(std::string_view name) {
  if (name == "core") return Profile::kCore;
  if (name == "integrity") return Profile::kIntegrity;
  if (name == "signed") return Profile::kSigned;
  if (name == "encrypted") return Profile::kEncrypted;
  if (name == "render") return Profile::kRender;
  return fail(ErrorCode::kUnsupported, "unknown profile: " + std::string(name));
}

std::string_view to_string(Profile profile) {
  switch (profile) {
    case Profile::kCore:      return "core";
    case Profile::kIntegrity: return "integrity";
    case Profile::kSigned:    return "signed";
    case Profile::kEncrypted: return "encrypted";
    case Profile::kRender:    return "render";
  }
  return "unknown";
}

Result<ValidationReport> validate(const Container& container,
                                  const Document& document, Profile profile) {
  ValidationReport r;
  r.profile = profile;

  check_core(container, document, r);
  if (profile_at_least(profile, Profile::kIntegrity)) {
    check_integrity(container, document, r);
  }
  if (profile_at_least(profile, Profile::kSigned)) {
    check_signed(container, r);
  }
  if (profile_at_least(profile, Profile::kEncrypted)) {
    check_encrypted(container, r);
  }
  if (profile_at_least(profile, Profile::kRender)) {
    check_render(container, document, r);
  }

  r.ok = r.issues.empty();
  return r;
}

}  // namespace mcdf

File: apps/web/packages/mcdf-ts/src/core/validate.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Sukesh Ashok Kumar and The MCDF Project
//
// Profile validation, ported from `src/core/validate.cpp`. The error codes are
// normative (conformance/errors.md) — messages are free-form, codes are not.

import { ContainerError, checkMemberPath, type Container } from '../container/container.js';
import { EncPublicKey } from '../crypto/enc-keys.js';
import { isAllowedHashAlgorithm } from '../crypto/hash.js';
import type { McdfDocument } from '../model/types.js';
import { ENCRYPTION_POLICY_PATH, parseEncryptionPolicyYaml } from '../serialize/policy.js';
import { verifyManifest } from './manifest.js';
import { RENDER_FORMATS, renderCanonical } from './render.js';
import { encryptedMembers, readEncryptionPolicy } from './sealed.js';
import { verifyContainer } from './sign.js';

export const PROFILES = ['core', 'integrity', 'signed', 'encrypted', 'render'] as const;

export type Profile = (typeof PROFILES)[number];

const PROFILE_RANK: Record<Profile, number> = {
  core: 0,
  integrity: 1,
  signed: 2,
  encrypted: 3,
  render: 4,
};

export function isProfile(name: string): name is Profile {
  return (PROFILES as readonly string[]).includes(name);
}

export interface ValidationIssue {
  code: string;
  message: string;
}

export interface ValidationReport {
  profile: Profile;
  ok: boolean;
  issues: ValidationIssue[];
}

function atLeast(profile: Profile, floor: Profile): boolean {
  return PROFILE_RANK[profile] >= PROFILE_RANK[floor];
}

function checkCore(container: Container, doc: McdfDocument, issues: ValidationIssue[]): void {
  if (!doc.hasContent) {
    issues.push({ code: 'E_MISSING_CONTENT', message: 'content.md is missing' });
  }

  for (const path of container.list()) {
    try {
      checkMemberPath(path);
    } catch (e) {
      const err = e as ContainerError;
      issues.push({ code: err.code ?? 'E_ASSET_PATH_ESCAPE', message: err.message });
    }
  }

  if (!doc.hasSchema) return; // no schema, no structural claims to check

  // The set of section ids that bind. Normally that comes from the headings; a
  // sealed content.md has none a validator can read, so spec §5.2.1 puts the
  // same information in the policy and the rule below does not change.
  const contentSealed = doc.sealed.includes('content.md');
  let bound: Set<string>;

  if (contentSealed) {
    const policy = readEncryptionPolicy(container);
    if (policy === null || !policy.structure.present) {
      // Without the attestation the binding is simply not knowable. Saying so
      // once is right; reporting every required section as "missing" would
      // blame the document for the encryptor's omission.
      issues.push({
        code: 'E_POLICY_INVALID',
        message:
          'content.md is sealed but encryption/policy.yaml carries no structure ' +
          'attestation, so schema binding cannot be evaluated (spec §5.2.1)',
      });
      return;
    }
    bound = new Set(policy.structure.bound_sections);
  } else {
    bound = new Set(doc.headings.map((h) => h.id).filter((id) => id !== ''));
  }

  const where = contentSealed ? 'in the sealed content' : 'in content';
  for (const s of doc.schema.sections) {
    if (s.id === '' || bound.has(s.id)) continue;
    issues.push(
      s.required
        ? {
            code: 'E_REQUIRED_SECTION_MISSING',
            message: `required section has no matching heading ${where}: ${s.id}`,
          }
        : {
            code: 'E_SCHEMA_UNBOUND',
            message: `schema section has no matching heading ${where}: ${s.id}`,
          },
    );
  }
}

function checkIntegrity(
  container: Container,
  doc: McdfDocument,
  issues: ValidationIssue[],
): void {
  if (!doc.hasManifest) {
    issues.push({ code: 'E_MISSING_MANIFEST', message: 'manifest.json is missing' });
    return;
  }
  if (!isAllowedHashAlgorithm(doc.manifest.hash_algorithm)) {
    issues.push({
      code: 'E_ALGO_NOT_ALLOWED',
      message: `hash algorithm not allowed: ${doc.manifest.hash_algorithm}`,
    });
    return;
  }

  const v = verifyManifest(container, doc.manifest);
  for (const p of v.missing) {
    issues.push({
      code: 'E_MANIFEST_MISSING_FILE',
      message: `file listed in manifest is missing: ${p}`,
    });
  }
  for (const p of v.mismatched) {
    issues.push({ code: 'E_MANIFEST_HASH_MISMATCH', message: `content hash mismatch: ${p}` });
  }
  for (const p of v.extra) {
    issues.push({
      code: 'E_MANIFEST_EXTRA_FILE',
      message: `file present but not listed in manifest: ${p}`,
    });
  }
}

/**
 * Every `signatures/*.sig` must verify over the canonical manifest.
 *
 * Each failure carries its own normative code out of the JWS verifier, so a
 * signature that is merely wrong stays distinguishable from one that could not
 * be evaluated — an unresolvable `kid` and a forged signature are different
 * problems with different fixes.
 */
function checkSigned(container: Container, issues: ValidationIssue[]): void {
  let checks;
  try {
    checks = verifyContainer(container);
  } catch (e) {
    // A missing manifest is already reported by the Integrity checks, which
    // always run first at this profile; reporting it twice would be noise.
    const code = (e as { code?: string }).code;
    if (code !== 'E_MISSING_MANIFEST') {
      issues.push({ code: code ?? 'E_IO', message: (e as Error).message });
    }
    return;
  }

  if (checks.length === 0) {
    issues.push({ code: 'E_SIG_MISSING', message: 'no signatures present (signatures/*.sig)' });
    return;
  }

  for (const check of checks) {
    if (check.valid) continue;
    issues.push({
      code: check.code ?? 'E_SIG_INVALID',
      message: `${check.file}: ${
        check.error === ''
          ? 'signature does not verify over the canonical manifest'
          : check.error
      }`,
    });
  }
}

/**
 * Static soundness of the encryption layer. No key is available here, so the
 * decrypt-time codes (E_NOT_A_RECIPIENT, E_DECRYPT_FAILED) stay out of scope;
 * a container with no policy has nothing to assert and passes.
 */
function checkEncrypted(container: Container, issues: ValidationIssue[]): void {
  if (!container.has(ENCRYPTION_POLICY_PATH)) return;

  let policy;
  try {
    policy = parseEncryptionPolicyYaml(container.readText(ENCRYPTION_POLICY_PATH));
  } catch (e) {
    issues.push({ code: 'E_POLICY_INVALID', message: (e as Error).message });
    return;
  }

  if (policy.method !== 'aes-256-gcm') {
    issues.push({
      code: 'E_ALGO_NOT_ALLOWED',
      message: `encryption method not allowed: ${policy.method}`,
    });
  }
  if (policy.key_management !== 'hpke') {
    issues.push({
      code: 'E_ALGO_NOT_ALLOWED',
      message: `key management not allowed: ${policy.key_management}`,
    });
  }
  if (policy.encrypted_files.length === 0) {
    issues.push({ code: 'E_POLICY_INVALID', message: 'policy lists no encrypted files' });
  }
  for (const file of policy.encrypted_files) {
    if (!container.has(file)) {
      issues.push({
        code: 'E_POLICY_INVALID',
        message: `encrypted file listed in policy is missing: ${file}`,
      });
    }
  }
  if (policy.recipients.length === 0) {
    issues.push({ code: 'E_POLICY_INVALID', message: 'policy has no recipients' });
  }

  // The attestation's *presence* rules (spec §5.2.1). The "required but absent"
  // case is reported by checkCore, which runs at every profile; what is left
  // here is the reverse — an attestation nobody can corroborate, which either
  // leaks the schema or claims something about content that is not sealed.
  if (policy.structure.present) {
    const sealed = encryptedMembers(container);
    if (!sealed.includes('content.md')) {
      issues.push({
        code: 'E_POLICY_INVALID',
        message: 'policy carries a structure attestation but content.md is not encrypted',
      });
    } else if (sealed.includes('schema.yaml') || !container.has('schema.yaml')) {
      issues.push({
        code: 'E_POLICY_INVALID',
        message:
          'policy carries a structure attestation with no readable schema.yaml to ' +
          'corroborate it (spec §5.2.1)',
      });
    }
  }
  for (const recipient of policy.recipients) {
    if (recipient.id === '' || recipient.enc === '' || recipient.wrapped_key === '') {
      issues.push({ code: 'E_POLICY_INVALID', message: 'recipient entry is incomplete' });
      continue;
    }
    try {
      EncPublicKey.fromDidKey(recipient.id);
    } catch {
      issues.push({
        code: 'E_KID_UNRESOLVABLE',
        message: `recipient id is not a resolvable X25519 did:key: ${recipient.id}`,
      });
    }
  }
}

/**
 * Both canonical renders must be producible — except over sealed content, which
 * cannot be rendered at all, so the profile is not applicable rather than failed
 * (spec §6).
 *
 * This used to report `E_UNIMPLEMENTED`: the render had to be byte-identical to
 * the reference and there was nothing published to prove that against. §10.4 and
 * the `canonical/render-*` vectors closed both halves, so the profile is claimed.
 */
function checkRender(container: Container, doc: McdfDocument, issues: ValidationIssue[]): void {
  if (doc.sealed.includes('content.md')) return;
  for (const format of RENDER_FORMATS) {
    try {
      renderCanonical(container, format);
    } catch (e) {
      issues.push({ code: 'E_IO', message: `render ${format} failed: ${(e as Error).message}` });
    }
  }
}

export function validate(
  container: Container,
  doc: McdfDocument,
  profile: Profile,
): ValidationReport {
  const issues: ValidationIssue[] = [];

  checkCore(container, doc, issues);
  if (atLeast(profile, 'integrity')) checkIntegrity(container, doc, issues);
  if (atLeast(profile, 'signed')) checkSigned(container, issues);
  if (atLeast(profile, 'encrypted')) checkEncrypted(container, issues);
  if (atLeast(profile, 'render')) checkRender(container, doc, issues);

  return { profile, ok: issues.length === 0, issues };
}
